### PR TITLE
pruntime: Implement dcap key handover

### PR DESCRIPTION
**Pull request by kvinwang**
This PR implements the key handover feature according to the design from @shelvenzhou.

the RTP time is got from at least two of the white listed servers as show below:
```rust
/// Ref: https://github.com/jauderho/nts-servers/
const TRUSTED_NTS_SERVERS: &[&str] = &[
    "time.cloudflare.com",
    "gps.ntp.br",
    "a.st1.ntp.br",
    "paris.time.system76.com",
    "ntp3.fau.de",
    "ptbtime1.ptb.de",
    "ntppool1.time.nl",
    "nts.netnod.se",
];
```

A few points to note:

- We use cfnts as nts client with much [refactor here](https://github.com/Phala-Network/phala-nts/pull/1) to make it meet our requirements.
- Because our phala-nts introduce dep of latest rustls, some of the crates in the deps tree conflicts. So we have to patch them.
- SGX Local attestation is used instead of the RA in the design, I think it meet the requirement.
- *Since RTP use trusted server from foreign countries and over UDP, it is hard to success get time from those servers in China mainland*.
- `dev_mode` is not fully supported in this version (Won't success in dev_mode). (Maybe added in the future).

### CODE REVIEW

File: crates/phactory/src/nts.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        .into_iter()
+        .filter_map(|r| r.ok())
+        .map(|r| r.receive_time_duration().as_secs())
+        .collect::<Vec<_>>();
+    info!("Got time from {} servers", results.len());
+    validate_results(results).context("Failed to get time from NTS servers")
+}
+
+fn validate_results(results: Vec<u64>) -> Result<u64> {
+    if results.len() < 2 {
```
Comment (shelvenzhou) [inline crates/phactory/src/nts.rs:35]: make these thresholds as constants so we can easily notice and update them

File: crates/phactory/src/nts.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        anyhow::bail!("Not enough results");
+    }
+    let average = results.iter().sum::<u64>() / results.len() as u64;
+    let max_diff = results
+        .iter()
+        .map(|r| (*r as i64 - average as i64).unsigned_abs())
+        .max()
+        .unwrap_or_default();
+    if max_diff > 60 {
```
Comment (shelvenzhou) [inline crates/phactory/src/nts.rs:44]: as above

File: crates/phactory/src/prpc_service.rs [hunk 3]
```diff
@@ -1807,6 +1839,213 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi for Rpc
...
+        };
+        sgx_api_lite::verify(recv_local_report).map_err(|_| from_display("No remote handover"))?;
+        let handler_hash = blake2_256(&request.encoded_challenge_handler);
+        if handler_hash != recv_local_report.body.report_data[..32] {
+            return Err(from_display("Invalid challenge handler"));
+        }
+        // 4. verify challenge block height and report timestamp
+        // only challenge within 150 blocks (30 minutes) is accepted
+        let challenge_height = challenge.block_number;
+        if !(challenge_height <= block_number && block_number - challenge_height <= 150) {
```
Comment (shelvenzhou) [inline crates/phactory/src/prpc_service.rs:1903]: This seems superceded by the `challenge.ntp_time_secs` check above

File: crates/phactory/src/prpc_service.rs [hunk 4]
```diff
@@ -1807,6 +1839,213 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi for Rpc
...
+        let mut report_data = [0u8; 64];
+        report_data[..32].copy_from_slice(&handler_hash);
+        // generate local attestation report to ensure the handover pRuntimes are on the same machine
+        let la_report = {
+            let its_target_info = unsafe {
+                sgx_api_lite::decode(&challenge.sgx_target_info)
+                    .map_err(|_| from_display("Invalid client sgx target info"))?
+            };
+            // the report data does not matter since we only care about the origin
+            let report = sgx_api_lite::report(its_target_info, &report_data)
```
Comment (shelvenzhou) [inline crates/phactory/src/prpc_service.rs:1993]: Does this mean an IAS pRuntime can also get key from DCAP pRuntime?

**Comment by shelvenzhou** [comment]
Double check @kvinwang : is NTP Authentication enabled in phala-nts?

**Comment by shelvenzhou** [comment]
I would also like to highlight some constants for future reference:
1. The handover is only allowed on the same physical machine, remote handover is not allowed;
2. The key sender must be a DCAP pRuntime, whether the receiver must be a DCAP is under checking;
3. Now we use NTP server for timestamp. We have 8 time servers, and will request all of them; **at least 2 responses** need to be received within **5s timeout**, and their difference need to be **within 60s**. The **average** is used as final timestamp;
4. For receiver worker: its latest block must be **within one week** of the challenge timestamp, and it must reply to the challenge **within 1min**.

**Comment by nanometerzhu** [comment]
Regarding to the NTP servers, there are 5 servers are in north Europe, and 2 are in Brazil. Are they too close?

File: crates/phactory/src/prpc_service.rs (hunk 4 above)
Comment (kvinwang) [reply crates/phactory/src/prpc_service.rs:1993]: No, IAS pruntime validates RA to each other during handover, while dcap pruntime server won't generate RA report. So, an IAS pruntime won't accept an handover from DCAP pruntime.

**Comment by kvinwang** [comment]
> is NTP Authentication enabled in phala-nts?

Yes, it's enabled.

**Comment by kvinwang** [comment]
> Regarding to the NTP servers, there are 5 servers are in north Europe, and 2 are in Brazil. Are they too close?

@shelvenzhou do you think this is a problem?

File: crates/phactory/src/prpc_service.rs (hunk 3 above)
Comment (kvinwang) [reply crates/phactory/src/prpc_service.rs]: Removed this one.

File: crates/phactory/src/nts.rs (hunk 2 above)
Comment (kvinwang) [reply crates/phactory/src/nts.rs]: Done

File: crates/phactory/src/nts.rs (hunk 1 above)
Comment (kvinwang) [reply crates/phactory/src/nts.rs]: Done

**Comment by shelvenzhou** [comment]
> > Regarding to the NTP servers, there are 5 servers are in north Europe, and 2 are in Brazil. Are they too close?
> 
> @shelvenzhou do you think this is a problem?

Can we make this configurable with on-chain governance?

**Comment by kvinwang** [comment]
> Can we make this configurable with on-chain governance?

Good idea.

**Review by shelvenzhou — APPROVED** [review]
LGTM

**Pull request merged**